### PR TITLE
Layout get view at

### PR DESCRIPTION
**Pull request by JJones780**
_Note: these changes require a re-run of `cmake build`._ 

I had promised myself to have a look at getViewAt with regard to upgrading the linear search put in for the multilayout merge. This is the result.

Please have a look at these changes and let me know if they are welcome.
 

The Layout class already contained  ( and in some cases knew then forgot ) all that was required for the most efficient getViewAt function; so that's where it now resides.
```
Layout.h:
XojPageView* getViewAt(int x, int y);
```

the sizeRow and sizeCol are now persist as members of Layout:
```
std::vector<int> sizeCol;
std::vector<int> sizeRow;
```

I tested out a bunch of searches and decided on a linear search up or down from the last result ( individually for row and column ) was the best option.  The binary search was actually the slowest of those I tried but I've left it  in (commented out) as it is handy to use for debugging and someone might want to see for themselves. The test with a bunch of searches and output of results to stderr is at checkin "also compare ..."  cbe0ebd.  I can also provide pdf's of graphs. Yes... I feel I overdid this.

The **pageposition, pagepositionHandler and pagePositionCache** classes are no longer used **and have been removed**.

### CODE REVIEW

File: src/gui/LayoutMapper.cpp [hunk 1]
```diff
@@ -55,8 +35,13 @@ LayoutMapper::LayoutMapper(int numPages, Settings* settings)
...
 
-	LayoutType type = compileLayout(isVertical, isRightToLeft, isBottomToTop);
-	layoutMapperInit(pages, numRows, numCols, fixRows, type, isPairedPages, pairsOffset);
+	//assemble bitflags for LayoutType
+	int type = isVertical ? LayoutBitFlags::Vertically : 0;
+	type |= isRightToLeft ? LayoutBitFlags::RightToLeft : 0;
+	type |= isBottomToTop ? LayoutBitFlags::BottomToTop : 0;
+
+	
+	layoutMapperInit(pages, numRows, numCols, fixRows, (LayoutType)type, isPairedPages, pairsOffset);
```
Comment (LittleHuba) [inline src/gui/LayoutMapper.cpp]: You could merge `configureFromSettings()` and `layoutMapperInit()` as this is the only position where you call the method. But depending on what you want to add you can also leave those two methods separate. This is no real issue, just something you may think about.

File: src/gui/Layout.cpp [hunk 2]
```diff
@@ -414,4 +424,97 @@ void Layout::ensureRectIsVisible(int x, int y, int width, int height)
...
+	{
+		for (; c < numCols; c++)
+		{
+			if (x <=   this->sizeCol[c]) break;
+		}
+		
+	}
+	
+	
+	if (c < numCols && r < numRows)
```
[the remark this answers is not in the document]
Comment (andreasb242) [reply src/gui/Layout.cpp]: And if there is one, probably a g_warning should be printed...

File: src/gui/LayoutMapper.cpp [hunk 3]
```diff
@@ -1,42 +1,22 @@
...
-{
-
+LayoutMapper::LayoutMapper(){
+	
 	XOJ_INIT_TYPE(LayoutMapper);
-
-	layoutMapperInit(pages, numRows, numCols, useRows, type, isPaired, firstPageOffset);
-
+	
+	this->actualPages = 0;
```
Comment (andreasb242) [inline src/gui/LayoutMapper.cpp]: We use a really recent C++ Version, which allow to initialize vars direct in the header file, really makes sense I think, even if it's not yet that common in C++.

This is not a problem or so, only a tip ;-)

File: src/gui/Layout.cpp [hunk 4]
```diff
@@ -414,4 +424,97 @@ void Layout::ensureRectIsVisible(int x, int y, int width, int height)
...
+
+
+						
+	/* Linear Up or Down Search from last position: */
+	
+	// Rows:
+	r = MAX(0, this->lastGetViewAtRow - 1);
+	if (r > 0 && y <= this->sizeRow[r]) //search lower
+	{
+		for (r--; r>=0; r--)
```
[the remark this answers is not in the document]
Comment (JJones780) [reply src/gui/Layout.cpp]: I did try that.. but no, it's currently not skipping the cached row.
The reason is that the cached page check is different than the grid row,column search. The page check doesn't include the blank area around each page.  
One could make the initial 'cache check' on the row,column boundaries (and skip it during the linear search) but that takes about as much overhead as letting the linear search do it (as that's where it starts).
  I considered taking out the initial cached page check altogether. Now that I look at it I see that when the user is drawing on the page this function is not being called... so it's not any more of a priority and should probably e removed. just cache the row and column to start searching from.

File: src/gui/Layout.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (JJones780) [reply src/gui/Layout.cpp]: This is an expected occurrence when the user is zoomed out and the pointer goes past the right or bottom edge ( and thus search increments past sizeCol, sizeRow arrays ).

File: src/gui/LayoutMapper.cpp (hunk 1 above)
Comment (JJones780) [reply src/gui/LayoutMapper.cpp]: That could probably be cleaned up a bit more now that it's settled in. Probably best in another commit so I'll have to come back to it.

File: src/gui/LayoutMapper.cpp (hunk 3 above)
Comment (JJones780) [reply src/gui/LayoutMapper.cpp]: Thanks,  Keep 'em coming!

**Comment by JJones780** [comment]
Thanks @LittleHuba and @andreasb242 .
I appreciate you both taking the time to review my changes. 
I'll update this pull request soon with your suggestions.

File: src/gui/Layout.cpp [hunk 5]
```diff
@@ -301,7 +305,7 @@ void Layout::layoutPages()
 					int paddingLeft;
 					int paddingRight;
 					int columnPadding = 0;
-					columnPadding = (sizeCol[c] - vDisplayWidth);
+					columnPadding = (this->sizeCol[c] - vDisplayWidth);
```
[the remark this answers is not in the document]
Comment (JJones780) [reply src/gui/Layout.cpp]: oops.. and the redundant line above it ;) fixed.

File: src/gui/Layout.cpp [hunk 6]
```diff
@@ -414,4 +424,97 @@ void Layout::ensureRectIsVisible(int x, int y, int width, int height)
...
+	
+	XOJ_CHECK_TYPE(Layout);
+
+	//try cached result first
+	if (this->lastGetViewAtPageNum < this->view->viewPagesLen && this->view->viewPages[this->lastGetViewAtPageNum]->containsPoint(x,y,false))
+	{
+		return this->view->viewPages[this->lastGetViewAtPageNum];
+	}
+	
+	int r;
```
[the remark this answers is not in the document]
Comment (JJones780) [reply src/gui/Layout.cpp]: done

**Comment by JJones780** [comment]
I removed the initial  cached page index check. A cache hit saves 200ns but a miss adds 4000ns. This function isn't called during a scribble so that 200ns isn't imperative. 
Made changes suggested by review - thanks.

**Pull request merged**